File: PlantID/Repositories/PlantRepository.swift
import Foundation

protocol PlantRepository {
    func insert(_ plant: Plant) throws
    func update() throws
    func delete(_ plant: Plant) throws

    func getAlivePlants() throws -> [Plant]
    func getArchivedPlants() throws -> [Plant]
    func getPlantById(_ id: UUID) throws -> Plant?
    func getPlantByNfcTag(_ tagId: String) throws -> Plant?
    func getAllNfcTagIds() throws -> [String]
    func getAllAlivePlantsSnapshot() throws -> [Plant]
    func countAlivePlants() throws -> Int
}

File: app/src/main/java/com/example/plant_id/data/repository/PlantRepository.kt
package com.example.plant_id.data.repository

import com.example.plant_id.data.dao.PlantDao
import com.example.plant_id.data.entity.Plant
import kotlinx.coroutines.flow.Flow

class PlantRepository(private val plantDao: PlantDao) {

    suspend fun insert(plant: Plant): Long = plantDao.insert(plant)

    suspend fun update(plant: Plant) = plantDao.update(plant)

    suspend fun delete(plant: Plant) = plantDao.delete(plant)

    fun getAlivePlants(): Flow<List<Plant>> = plantDao.getAlivePlants()

    fun getArchivedPlants(): Flow<List<Plant>> = plantDao.getArchivedPlants()

    fun getPlantById(id: Long): Flow<Plant?> = plantDao.getPlantById(id)

    suspend fun getPlantByNfcTag(tagId: String): Plant? = plantDao.getPlantByNfcTag(tagId)

    suspend fun getAllNfcTagIds(): List<String> = plantDao.getAllNfcTagIds()

    suspend fun getAllAlivePlantsSnapshot(): List<Plant> = plantDao.getAllAlivePlantsSnapshot()

    suspend fun countAlivePlants(): Int = plantDao.countAlivePlants()
}
